tests/model_fields/test_jsonfield.py

```diff
@@ -0,0 +1,710 @@
+            with self.subTest(value=value):
+                obj = JSONModel.objects.create(value=value)
+                obj.refresh_from_db()
+                self.assertEqual(obj.value, value)
+
+    def test_list(self):
+        values = [
+            [],
+            ['John', 20, 180.3],
+            [True, [False, None]],
+        ]
+        for value in values:
+            with self.subTest(value=value):
+                obj = JSONModel.objects.create(value=value)
+                obj.refresh_from_db()
+                self.assertEqual(obj.value, value)
+
+    def test_realistic_object(self):
+        value = {
+            'name': 'John',
+            'age': 20,
+            'pets': [
+                {'name': 'Kit', 'type': 'cat', 'age': 2},
+                {'name': 'Max', 'type': 'dog', 'age': 1},
+            ],
+            'courses': [
+                ['A1', 'A2', 'A3'],
+                ['B1', 'B2'],
+                ['C1'],
+            ],
+        }
+        obj = JSONModel.objects.create(value=value)
+        obj.refresh_from_db()
+        self.assertEqual(obj.value, value)
+
+
+@skipUnlessDBFeature('supports_json_field')
+class TestQuerying(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        cls.primitives = [True, False, 'yes', 7, 9.6]
+        values = [
+            None,
+            [],
+            {},
+            {'a': 'b', 'c': 14},
+            {
+                'a': 'b',
+                'c': 14,
+                'd': ['e', {'f': 'g'}],
+                'h': True,
+                'i': False,
+                'j': None,
+                'k': {'l': 'm'},
+                'n': [None],
+            },
+            [1, [2]],
+            {'k': True, 'l': False},
+            {
+                'foo': 'bar',
+                'baz': {'a': 'b', 'c': 'd'},
+                'bar': ['foo', 'bar'],
+                'bax': {'foo': 'bar'},
+            },
+        ]
+        cls.objs = [
+            NullableJSONModel.objects.create(value=value)
+            for value in values
+        ]
+        if connection.features.supports_primitives_in_json_field:
+            cls.objs.extend([
+                NullableJSONModel.objects.create(value=value)
+                for value in cls.primitives
+            ])
+
+    def test_has_key(self):
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__has_key='a'),
+            [self.objs[3], self.objs[4]],
+        )
+
+    def test_has_key_null_value(self):
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__has_key='j'),
+            [self.objs[4]],
+        )
+
+    def test_has_key_list(self):
+        obj = NullableJSONModel.objects.create(value=[{'a': 1}, {'b': 'x'}])
+        tests = [
+            Q(value__0__has_key='a'),
+            Q(value__has_key=KeyTransform('a', KeyTransform(0, 'value'))),
+            Q(value__has_key=KeyTransform('a', KeyTransform('0', 'value'))),
```

Comment:
Cases from lines 361 and 363 work with the previous implementation.

Reply:
Are you sure about that? I just tried it from the latest `master` and added the following test case (basically the same test but with the model adjusted with `JSONModel`):

```py
    def test_has_key_list(self):
        obj = JSONModel.objects.create(field=[{'a': 1}, {'b': 'x'}])
        tests = [
            Q(field__0__has_key='a'),
            Q(field__has_key=KeyTransform('a', KeyTransform(0, 'field'))),
            Q(field__has_key=KeyTransform('a', KeyTransform('0', 'field'))),
        ]
        for condition in tests:
            with self.subTest(condition=condition):
                self.assertSequenceEqual(
                    JSONModel.objects.filter(condition),
                    [obj],
                )
```

The `Q(field__has_key=KeyTransform('a', KeyTransform(0, 'field')))` test failed with:

```
django.db.utils.DataError: invalid input syntax for integer: "a"
LINE 1: ...ield" ? ("postgres_tests_jsonmodel"."field" #> ARRAY[0,'a'])
                                                                  ^
```

While `Q(value__has_key=KeyTransform('a', KeyTransform('0', 'value')))` failed with:

```
django.db.utils.ProgrammingError: operator does not exist: jsonb ? jsonb
LINE 1: ...onmodel" WHERE "postgres_tests_jsonmodel"."field" ? ("postgr...
                                                             ^
HINT:  No operator matches the given name and argument type(s). You might need to add explicit type casts.
```

Reply:
Also: case from line 361 still works with the new implementation. It's 362 and 363 that don't, which seem to be the case with the previous implementation as well (albeit the new implementation throws up the same error for 362 as with 363, i.e. `jsonb ? jsonb`).

Unless I'm missing something...

Reply:
True, I mixed up sth when switching branches. Nevertheless they work on all databases except PostgreSQL, so we need to fix this.